jparest/src/main/java/uk/gov/homeoffice/digital/sas/jparest/repository/TenantRepositoryImpl.java

```diff
@@ -12,11 +37,177 @@
  * Resources for ManyToMany relationships can also be queried.
  */
 public class TenantRepositoryImpl<T>
-    extends SimpleJpaRepository<T, UUID> implements TenantRepository<T> {
+        extends SimpleJpaRepository<T, UUID> implements TenantRepository<T> {
+
+  private final EntityManager entityManager;
+  private final Class<T> entityType;
+  private final String tenantIdFieldName;
+  private final PersistenceUnitUtil persistenceUnitUtil;
+
+  private static final String QUERY_HINT = "javax.persistence.fetchgraph";
+
 
   public TenantRepositoryImpl(Class<T> entityType, EntityManager entityManager) {
     super(entityType, entityManager);
+    this.entityManager = entityManager;
+    this.entityType = entityType;
+    this.tenantIdFieldName = getTenantIdEntityFieldName();
+    this.persistenceUnitUtil = entityManager.getEntityManagerFactory().getPersistenceUnitUtil();
+  }
+
+
+  @Override
+  @Transactional
+  public List<T> findAllByTenantId(UUID tenantId, SpelExpression filter, Pageable pageable) {
+
+    CriteriaBuilder builder = this.entityManager.getCriteriaBuilder();
+    CriteriaQuery<T> query = builder.createQuery(entityType);
+    Root<T> root = query.from(entityType);
+
+    Predicate tenantPredicate = builder.equal(root.get(tenantIdFieldName), tenantId);
+    Predicate filterPredicate = SpelExpressionToPredicateConverter.convert(filter, builder, root);
+    Predicate finalPredicate =
+        filter != null ? builder.and(tenantPredicate, filterPredicate) : tenantPredicate;
+    query.where(finalPredicate);
+
+    CriteriaQuery<T> select = query.select(root);
+    select.orderBy(getOrderCriteria(pageable.getSort(), root, builder));
+
+    return this.entityManager.createQuery(select)
+        .setFirstResult((int) pageable.getOffset())
+        .setMaxResults(pageable.getPageSize())
+        .setHint(QUERY_HINT, entityManager.createEntityGraph(entityType))
+        .getResultList();
+  }
+
+
+  @Override
+  @Transactional
```

Comment:
I think a good practice in relation to transaction management is to leave it to the service layer to drive. For example if a service method's intention is to perform multiple database operations in the same transaction.
https://stackoverflow.com/a/1079125/2219361

Reply:
Ah so my thinking was based on the fact that spring's `SimpleJpaRepository` has the entire class marked up as `@Transactional` so I thought given we're extending that our methods should also be.
But yes I agree the service layer should have main control over the transactions with the annotation to come as part of the next subtask eahw-2545.

So would you say given the `SimpleJpaRepository` example, it's still unnecessary to have our repo methods marked up with `transactional` given we're controlling it in the service?

Reply:
I've removed the transactional annotations from our repo implementation methods. As discussed, in summary our transaction controlling in the service is sufficient and in fact probably better to have that be the only source of transaction management.

Reply:
As discussed with @daleelR , the transactional annotation is not necessary in this repository as transaction is currently explicitly managed by Service class.
Going forward, and part of next sub task (eahw-2545), transactional annotation will be placed in Service class